src/devices/cpu/tlcs900/tmp94c241.cpp

```diff
@@ -0,0 +1,1935 @@
+{
+	if (BIT(offset, 0))
+		m_block_cs[5] = (m_block_cs[5] & 0xff) | (data << 8);
+	else
+		m_block_cs[5] = (m_block_cs[5] & 0xff00) | data;
+}
+
+void tmp94c241_device::mamr0_w(offs_t offset, uint8_t data)
+{
+	SFR_MAMR0 = data;
+}
+
+void tmp94c241_device::msar0_w(offs_t offset, uint8_t data)
+{
+	SFR_MSAR0 = data;
+}
+
+uint8_t tmp94c241_device::mamr0_r()
+{
+	return SFR_MAMR0;
+}
+
+uint8_t tmp94c241_device::msar0_r()
+{
+	return SFR_MSAR0;
+}
+
+void tmp94c241_device::mamr1_w(offs_t offset, uint8_t data)
+{
+	SFR_MAMR1 = data;
+}
+
+void tmp94c241_device::msar1_w(offs_t offset, uint8_t data)
+{
+	SFR_MSAR1 = data;
+}
+
+uint8_t tmp94c241_device::mamr1_r()
+{
+	return SFR_MAMR1;
+}
+
+uint8_t tmp94c241_device::msar1_r()
+{
+	return SFR_MSAR1;
+}
+
+void tmp94c241_device::mamr2_w(offs_t offset, uint8_t data)
+{
+	SFR_MAMR2 = data;
+}
+
+void tmp94c241_device::msar2_w(offs_t offset, uint8_t data)
+{
+	SFR_MSAR2 = data;
+}
+
+uint8_t tmp94c241_device::mamr2_r()
+{
+	return SFR_MAMR2;
+}
+
+uint8_t tmp94c241_device::msar2_r()
+{
+	return SFR_MSAR2;
+}
+
+void tmp94c241_device::mamr3_w(offs_t offset, uint8_t data)
+{
+	SFR_MAMR3 = data;
+}
+
+void tmp94c241_device::msar3_w(offs_t offset, uint8_t data)
+{
+	SFR_MSAR3 = data;
+}
+
+uint8_t tmp94c241_device::mamr3_r()
+{
+	return SFR_MAMR3;
+}
+
+uint8_t tmp94c241_device::msar3_r()
+{
+	return SFR_MSAR3;
+}
+
+void tmp94c241_device::mamr4_w(offs_t offset, uint8_t data)
+{
+	SFR_MAMR4 = data;
+}
+
+void tmp94c241_device::msar4_w(offs_t offset, uint8_t data)
+{
+	SFR_MSAR4 = data;
+}
+
+uint8_t tmp94c241_device::mamr4_r()
+{
+	return SFR_MAMR4;
+}
+
+uint8_t tmp94c241_device::msar4_r()
+{
+	return SFR_MSAR4;
+}
+
+void tmp94c241_device::mamr5_w(offs_t offset, uint8_t data)
+{
+	SFR_MAMR5 = data;
+}
+
+void tmp94c241_device::msar5_w(offs_t offset, uint8_t data)
+{
+	SFR_MSAR5 = data;
+}
+
+uint8_t tmp94c241_device::mamr5_r()
+{
+	return SFR_MAMR5;
+}
+
+uint8_t tmp94c241_device::msar5_r()
+{
+	return SFR_MSAR5;
+}
+
+uint8_t tmp94c241_device::t8run_r()
+{
+	return SFR_T8RUN;
+}
+
+void tmp94c241_device::t8run_w(uint8_t data)
+{
+	SFR_T8RUN = data;
+
+	if ( !TIMER_0_IS_RUNNING )
+	{
+		UPCOUNTER_0 = 0;
+		TIMER_CHANGE_0 = 0;
+	}
+	if ( !TIMER_1_IS_RUNNING )
+	{
+		UPCOUNTER_1 = 0;
+		TIMER_CHANGE_1 = 0;
+	}
+	if ( !TIMER_2_IS_RUNNING )
+	{
+		UPCOUNTER_2 = 0;
+		TIMER_CHANGE_2 = 0;
+	}
+	if ( !TIMER_3_IS_RUNNING )
+	{
+		UPCOUNTER_3 = 0;
+		TIMER_CHANGE_3 = 0;
+	}
+}
+
+void tmp94c241_device::treg01_w(offs_t offset, uint8_t data)
+{
+
+    if (BIT(offset, 0))
+		SFR_TREG1 = data;
+	else
+		SFR_TREG0 = data;
+}
+
+uint8_t tmp94c241_device::t01mod_r()
+{
+	return SFR_T01MOD;
+}
+
+void tmp94c241_device::t01mod_w(uint8_t data)
+{
+	logerror("T01MOD = %02X\n", data);
+	SFR_T01MOD = data;
+}
+
+uint8_t tmp94c241_device::t02ffcr_r()
+{
+	return SFR_T02FFCR;
+}
+
+enum
+{
+	FF_INVERT,
+	FF_SET,
+	FF_CLEAR,
+	FF_DONTCARE
+};
+
+void tmp94c241_device::change_timer_flipflop(uint8_t flipflop, uint8_t operation)
```

Comment:
That method makes no sense.  Look at function 4 on PORT_D, INVERT clears bit 0 (never sets it), SET clears bit 1-7 if bit0=1, CLEAR sets bits 1-7 if bit0=1.  If it's the expected semantics (which I doubt), the names make no sense whatsoever.

Plus it feels like it's always use with constant parameters, which select which line to activate.  Is there a real reason to group them together rather than briging each line up at the calling point?

Reply:
I've just pushed a re-implementation of this method that I consider to be much better now.